File: test/gates/stride.test.ts
import { describe, expect, it } from 'vitest';
import {
  evaluateStride,
  parseStrideAllowlist,
  pathMatchesAllowlist,
} from '../../src/gates/stride.js';

/**
 * Build a synthetic unified diff around the given added/removed lines.
 * No network, no LLM: string fixtures only.
 */
function diffFor(file: string, added: string[], removed: string[] = []): string {
  const lines = [
    `diff --git a/${file} b/${file}`,
    `--- a/${file}`,
    `+++ b/${file}`,
    `@@ -1,${Math.max(removed.length, 1)} +1,${Math.max(added.length, 1)} @@`,
    ...removed.map((l) => `-${l}`),
    ...added.map((l) => `+${l}`),
  ];
  return lines.join('\n');
}

// Synthetic credential fixtures, assembled at runtime so this file never
// contains a usable credential-shaped literal (the STRIDE detector under
// test still sees the exact same string).
const CRED_LINE = 'const api_key = "sk-live-' + 'abcd1234";';
const CRED_LINE_ALT = 'const api_key = "sk-live-' + 'src-9999";';

describe('evaluateStride (G5 gate executor)', () => {
  const positives: Array<{ file: string; added?: string[]; removed?: string[]; category: string }> = [
    {
      file: 'src/spoof.ts',
      added: [CRED_LINE],
      category: 'Spoofing',
    },
    {
      file: 'src/tamper.ts',
      added: ['db.query(`SELECT * FROM u WHERE id = ${req.params.id}`)'],
      category: 'Tampering',
    },
    {
      file: 'src/repudiate.ts',
      added: ['// audit log call commented out'],
      category: 'Repudiation',
    },
    { file: 'src/disclose.ts', added: ['console.log(req.body)'], category: 'InformationDisclosure' },
    { file: 'src/dos.ts', added: ['setInterval(tick, 1000)'], category: 'DenialOfService' },
    {
      file: 'src/eop.ts',
      added: [],
      removed: ['if (user.role !== "admin") return;'],
      category: 'ElevationOfPrivilege',
    },
  ];

  const negatives: Array<{ file: string; added: string[]; category: string }> = [
    { file: 'src/spoof.ts', added: ['const config = loadConfig(env);'], category: 'Spoofing' },
    { file: 'src/tamper.ts', added: ['db.query("SELECT * FROM u WHERE id = ?", [id])'], category: 'Tampering' },
    { file: 'src/repudiate.ts', added: ['audit("user.updated", ctx);'], category: 'Repudiation' },
    { file: 'src/disclose.ts', added: ['logger.info("handled request");'], category: 'InformationDisclosure' },
    { file: 'src/dos.ts', added: ['await tick();'], category: 'DenialOfService' },
    { file: 'src/eop.ts', added: ['requireRole("admin");'], category: 'ElevationOfPrivilege' },
  ];

  // (a) one positive fixture per STRIDE category
  for (const p of positives) {
    it(`reports ${p.category} for its trigger shape`, async () => {
      const r = await evaluateStride({ diff: diffFor(p.file, p.added ?? [], p.removed ?? []) });
      expect(r.findings.some((f) => f.category === p.category)).toBe(true);
    });
  }

  // (b) one negative fixture per STRIDE category
  for (const n of negatives) {
    it(`reports no ${n.category} for the benign shape`, async () => {
      const r = await evaluateStride({ diff: diffFor(n.file, n.added) });
      expect(r.findings.some((f) => f.category === n.category)).toBe(false);
    });
  }

  it('(c) promotes a HIGH credential literal to CRITICAL', async () => {
    const r = await evaluateStride({ diff: diffFor('src/cred.ts', [CRED_LINE]) });
    const f = r.findings[0];
    expect(f).toBeDefined();
    expect(f?.severity).toBe('CRITICAL');
    expect(r.severityMax).toBe('CRITICAL');
  });

  it('(d) blocks on HIGH (api_key) with severityMax HIGH', async () => {
    const r = await evaluateStride({ diff: diffFor('src/key.ts', ['const api_key = process.env.API_KEY;']) });
    expect(r.findings[0]?.severity).toBe('HIGH');
    expect(r.severityMax).toBe('HIGH');
    expect(r.severityMax).not.toBe('CRITICAL');
  });

  it('(e) treats MEDIUM as advisory: severityMax MEDIUM, no HIGH/CRITICAL', async () => {
    const r = await evaluateStride({ diff: diffFor('src/adv.ts', ['console.log("debug:", password);']) });
    expect(r.severityMax).toBe('MEDIUM');
    expect(r.findings.every((f) => f.severity !== 'HIGH' && f.severity !== 'CRITICAL')).toBe(true);
  });

  it('(f) passes on empty diff with zero findings', async () => {
    for (const d of ['', '   ']) {
      const r = await evaluateStride({ diff: d });
      expect(r.findings).toEqual([]);
      expect(r.severityMax).toBeNull();
    }
  });

  it('(g) passes on null/undefined diff without throwing', async () => {
    for (const d of [null, undefined] as Array<string | null | undefined>) {
      const r = await evaluateStride({ diff: d as string });
      expect(r.findings).toEqual([]);
      expect(r.severityMax).toBeNull();
    }
  });

  it('(h) passes through contextDigest verbatim', async () => {
    const digest = 'sha256:deadbeef';
    const r = await evaluateStride({ diff: diffFor('src/x.ts', ['plain line']), contextDigest: digest });
    expect(r.contextDigest).toBe(digest);
  });

  it('(i) includes file and line provenance on findings', async () => {
    const r = await evaluateStride({ diff: diffFor('src/i.ts', ['console.log(req.body)']) });
    expect(r.findings[0]?.file).toBe('src/i.ts');
    expect(r.findings[0]?.line).toBe(1);
  });

  it('(j) suppresses findings whose file matches a committed allowlist path (PRD Q25)', async () => {
    // Fixture credential in a test file: HIGH + CRITICAL promotion without the allowlist…
    const diff = diffFor('test/fixtures/credentials.json', [CRED_LINE]);
    const blocked = await evaluateStride({ diff });
    expect(blocked.severityMax).toBe('CRITICAL');

    // …and fully suppressed once the PR carries the allowlist entry.
    const allowed = await evaluateStride({ diff, allowlistPaths: ['test/fixtures/**'] });
    expect(allowed.findings).toEqual([]);
    expect(allowed.severityMax).toBeNull();
  });

  it('(k) does not suppress findings outside the allowlist paths', async () => {
    const diff = [
      diffFor('test/fixtures/credentials.json', [CRED_LINE]),
      diffFor('src/cred.ts', [CRED_LINE_ALT]),
    ].join('\n');
    const r = await evaluateStride({ diff, allowlistPaths: ['test/fixtures/**'] });
    expect(r.findings.map((f) => f.file)).toEqual(['src/cred.ts']);
    expect(r.severityMax).toBe('CRITICAL');
  });

  it('(l) treats a missing/empty allowlist as no suppression', async () => {
    const diff = diffFor('src/key.ts', [CRED_LINE]);
    for (const allowlistPaths of [undefined, []]) {
      const r = await evaluateStride({ diff, allowlistPaths });
      expect(r.severityMax).toBe('CRITICAL');
    }
  });

  it('(m) treats a malformed allowlist as fail-closed (no suppression)', async () => {
    for (const text of ['not json', '["src/**"]', '{"paths": "src/**"}', '{"paths": [1]}', 'null']) {
      expect(parseStrideAllowlist(text)).toBeNull();
    }
    const diff = diffFor('src/key.ts', [CRED_LINE]);
    const parsed = parseStrideAllowlist('{"paths": [1]}');
    const r = await evaluateStride({ diff, allowlistPaths: parsed ?? undefined });
    expect(r.severityMax).toBe('CRITICAL');
  });

  it('(n) parses a well-formed allowlist into its path patterns', () => {
    expect(parseStrideAllowlist('{"paths": ["test/**", "fixtures/*.json"]}')).toEqual([
      'test/**',
      'fixtures/*.json',
    ]);
    expect(parseStrideAllowlist('')).toBeNull();
  });

  it('(o) matches allowlist globs: ** crosses slashes, * stays in segment, bare name matches basename', () => {
    const patterns = ['test/**', 'src/fixtures/*.json', 'golden.pem'];
    expect(pathMatchesAllowlist('test/fixtures/creds.json', patterns)).toBe(true);
    expect(pathMatchesAllowlist('src/fixtures/creds.json', patterns)).toBe(true);
    expect(pathMatchesAllowlist('src/fixtures/nested/creds.json', patterns)).toBe(false);
    expect(pathMatchesAllowlist('golden.pem', patterns)).toBe(true);
    expect(pathMatchesAllowlist('certs/golden.pem', patterns)).toBe(true);
    expect(pathMatchesAllowlist('src/creds.json', patterns)).toBe(false);
  });
});

File: internal/gates/stride_test.go
package gates

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func hunk(file string, added, removed []string, line *int) parsedDiffHunk {
	join := func(lines []string) string {
		if len(lines) == 0 {
			return ""
		}
		return strings.Join(lines, "\n") + "\n"
	}
	h := parsedDiffHunk{File: file, Added: join(added), Removed: join(removed), Line: line}
	return h
}

func linePtr(n int) *int { return &n }

// Synthetic credential fixture, assembled at runtime so this file never
// contains a usable credential-shaped literal (the gate under test still
// sees the exact same string).
const credLine = "const api_key = " + `"sk-live-` + `abcd1234";`

// mustRead reads a testdata fixture relative to this package.
func mustRead(t *testing.T, rel string) string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("testdata", rel))
	if err != nil {
		t.Fatalf("read testdata %s: %v", rel, err)
	}
	return string(data)
}

func TestRunStrideGateBlocksHardcodedAPIKey(t *testing.T) {
	r := RunStrideGate([]parsedDiffHunk{hunk("src/handler.ts", []string{credLine}, nil, linePtr(10))}, "/tmp/worktree")
	if r.Passed {
		t.Fatalf("expected gate to block, got %+v", r)
	}
	if r.SeverityMax != StrideSeverityHigh {
		t.Errorf("severityMax = %q, want high", r.SeverityMax)
	}
	if len(r.Findings) == 0 || r.Findings[0].Category != "S" {
		t.Errorf("findings[0].category = %v, want S", r.Findings)
	}
	if r.Gate != "G5-stride" {
		t.Errorf("gate = %q, want G5-stride", r.Gate)
	}
}

func TestRunStrideGatePassesMediumAdvisory(t *testing.T) {
	r := RunStrideGate([]parsedDiffHunk{hunk("src/handler.ts", []string{`console.log("debug:", password);`}, nil, linePtr(20))}, "/tmp/worktree")
	if !r.Passed {
		t.Fatalf("expected advisory pass, got %+v", r)
	}
	if r.SeverityMax != StrideSeverityMedium {
		t.Errorf("severityMax = %q, want medium", r.SeverityMax)
	}
	if len(r.Findings) < 1 {
		t.Errorf("expected at least one finding, got %+v", r.Findings)
	}
	if !strings.Contains(r.Detail, "STRIDE") || !strings.Contains(r.Detail, "## STRIDE G5 findings") {
		t.Errorf("detail missing STRIDE block: %q", r.Detail)
	}
}

func TestRunStrideGatePassesOnEmptyInput(t *testing.T) {
	for _, parsed := range [][]parsedDiffHunk{nil, {}} {
		r := RunStrideGate(parsed, "/tmp/wt")
		if !r.Passed {
			t.Errorf("expected pass on %v", parsed)
		}
		if len(r.Findings) != 0 {
			t.Errorf("expected zero findings on %v, got %+v", parsed, r.Findings)
		}
		if r.SeverityMax != "" {
			t.Errorf("expected empty severityMax on %v, got %q", parsed, r.SeverityMax)
		}
	}
}

func TestRunStrideGateOneFindingPerStrideLetter(t *testing.T) {
	hunks := []parsedDiffHunk{
		hunk("src/s.ts", []string{credLine}, nil, linePtr(1)),
		hunk("src/t.ts", []string{"db.query(`SELECT * FROM u WHERE id = ${req.params.id}`)"}, nil, linePtr(2)),
		hunk("src/r.ts", []string{"// audit log call removed"}, nil, linePtr(3)),
		hunk("src/i.ts", []string{"console.log(req.body)"}, nil, linePtr(4)),
		hunk("src/d.ts", []string{"setInterval(tick, 1000)"}, nil, linePtr(5)),
		hunk("src/e.ts", nil, []string{`if (user.role !== "admin") return;`}, linePtr(6)),
	}
	r := RunStrideGate(hunks, "/tmp/worktree")
	seen := map[string]bool{}
	for _, f := range r.Findings {
		seen[f.Category] = true
	}
	for _, letter := range []string{"S", "T", "R", "I", "D", "E"} {
		if !seen[letter] {
			t.Errorf("missing category %s in %+v", letter, r.Findings)
		}
	}
}

func TestRunStrideGateDuplicateLinesDistinctIDs(t *testing.T) {
	line := `console.log("debug:", password);`
	r := RunStrideGate([]parsedDiffHunk{hunk("src/dup.ts", []string{line, line}, nil, linePtr(7))}, "/tmp/worktree")
	if len(r.Findings) != 2 {
		t.Fatalf("expected 2 findings, got %d", len(r.Findings))
	}
	if r.Findings[0].ID == r.Findings[1].ID {
		t.Errorf("duplicate ids: %q", r.Findings[0].ID)
	}
}

func TestParseUnifiedDiffSplitsHunks(t *testing.T) {
	// Recorded fixture diff with the credential line substituted at runtime.
	diff := mustRead(t, "diffs/blocked-apikey.diff")
	diff = strings.Replace(diff, "const api_key = process.env.API_KEY;", credLine, 1)
	hunks := ParseUnifiedDiff(diff)
	if len(hunks) != 1 {
		t.Fatalf("expected 1 hunk, got %d", len(hunks))
	}
	if hunks[0].File != "src/key.ts" {
		t.Errorf("file = %q", hunks[0].File)
	}
	if hunks[0].Line == nil || *hunks[0].Line != 1 {
		t.Errorf("line = %v, want 1", hunks[0].Line)
	}
	if hunks[0].Added != credLine+"\n" {
		t.Errorf("added = %q", hunks[0].Added)
	}
	if hunks[0].Removed != "const old = 1;\n" {
		t.Errorf("removed = %q", hunks[0].Removed)
	}
}
